CIP-????/README.md

```diff
@@ -0,0 +1,97 @@
+Empirically, many are capped at about 30, whereas if they were limited only by the amount of space in the transaction for inputs and outputs the limit would be hundreds.
+
+If we had arrays with constant time indexing, we could make this linear instead (we would still need to do a linear amount of work to create arrays for the transaction inputs and outputs from the lists in the script context).
+
+#### `Constr` arguments
+
+The `Data` type has a `Constr` alternative which is used for encoding datatype constructors. 
+This is used for encoding the script context, and is used by languages such as Aiken extensively for representing user-defined datatypes also.
+
+The fields of the constructor are encoded in a list; hence to access a particular field the compiled code needs to do a linear amount of work. 
+If the arguments to a `Constr` were an array, we could access the fields in constant time.
+
+### Maps
+
+#### Operations on `Value`
+
+The `Value` type is a nested map: it is a map from bytestrings (representing policy IDs) to maps from bytestrings (representing token names) to integers (representing quantities).
+Since map operations are currently linear, this means that even simple operations like checking whether one value is less than another can have quadratic cost.
+
+This would be much better if map operations were logarithmic cost.
+
+#### Indexing by party
+
+Many applications have a known set of participants identified by some bytestring, typically a public key. 
+It is therefore natural to store per-party state in a map indexed by the party identifier.
+
+Since map operations are currently linear, this is needlessly expensive and imposes a limit on the number of parties.
+
+## Goals
+
+1. Reduce the cost of operations on `Value` by a factor of 2-10
+2. Reduce the cost of a matching algorithm such that we can handle hundreds of matches for the same cost it currently takes to do 30.
+
+## Open questions
+
+- Can we implement a set/map data structure in Plutus Core code that has acceptable performance and doesn’t require too much size overhead?
+- Do we need generic maps or is a map-from-bytestring sufficient? What about map-from-integer?
+    - Generic maps are harder since we typically need to know how to order the key type
+- Is an array type useful even if it is immutable?
+    - We are unlikely to be able to offer mutable arrays
+- Are builtin data structures useful enough even if they can only contain builtin types?
+    - This would mean that complex data structures would have to be stored inside arrays as `Data`, rather than using Scott encoding or sum-of-products representation
+- Can we feasibly change the structure of the builtin `Data` type so that `Constr` arguments are in an array?
+    - We would need to retain both versions for backwards compatibility
+    
+## References
+
+- https://x.com/Quantumplation/status/1733298551571038338?s=20
```

Comment:
```suggestion
```
We don't have a section for this in the CPS template... to me the discussion at this link seems split between _Problem_ and _Use cases_, so could it be linked in one or both those places as appropriate?

Reply:
Maybe we should add a `References` section. It's pretty useful.

Reply:
Good suggestion, see [Issue #731](https://github.com/cardano-foundation/CIPs/issues/731).

Reply:
... specifically @michaelpj we are talking about restoring the removed `## References` section in https://github.com/cardano-foundation/CIPs/issues/731#issuecomment-1874421788 and with consensus will add it back in via https://github.com/cardano-foundation/CIPs/pull/730.

Reply:
I'm preparing a PR that will document optional sections like this in the CPS template the way we added them for CIPs in #730.  We should have added it there, but by the time that discussion concluded I couldn't recall any CPS examples.  I'll have that PR for CIP-9999 submitted (and maybe merged) before this in final review, and in any case I believe a `## References` section by community consensus by editor unanimity is OK.

BUT @michaelpj @Quantumplation I don't think this link belongs just hanging here.  People not present in that Twitter discussion will have difficulty pulling the relevant elements out of it... so can you
- add some introduction about what is at this link?
- then consider moving it to _Problem_ or _Goals_ according to what this discussion most appropriately defines?

Reply:
p.s. the official inclusion of _References_ in the CPS has been proposed in https://github.com/cardano-foundation/CIPs/pull/756.